src/internal/components/file-token-group/file-token.tsx

```diff
@@ -0,0 +1,192 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import React, { useRef, useState } from 'react';
+import clsx from 'clsx';
+
+import InternalBox from '../../../box/internal.js';
+import { FormFieldError, FormFieldWarning } from '../../../form-field/internal';
+import InternalSpaceBetween from '../../../space-between/internal.js';
+import InternalSpinner from '../../../spinner/internal.js';
+import DismissButton from '../../../token-group/dismiss-button';
+import { TokenGroupProps } from '../../../token-group/interfaces.js';
+import { BaseComponentProps } from '../../base-component/index.js';
+import { useUniqueId } from '../../hooks/use-unique-id';
+import Tooltip from '../tooltip/index';
+import * as defaultFormatters from './default-formatters.js';
+import { FileOptionThumbnail } from './thumbnail.js';
+
+import styles from './styles.css.js';
+import testUtilStyles from './test-classes/styles.css.js';
+
+export namespace FileTokenProps {
+  export interface I18nStrings {
+    removeFileAriaLabel: (fileIndex: number) => string;
+    errorIconAriaLabel?: string;
+    warningIconAriaLabel?: string;
+    formatFileSize?: (sizeInBytes: number) => string;
+    formatFileLastModified?: (date: Date) => string;
+  }
+}
+
+export interface FileTokenProps extends BaseComponentProps {
+  file: File;
+  onDismiss: () => void;
+  showFileSize?: boolean;
+  showFileLastModified?: boolean;
+  showFileThumbnail?: boolean;
+  errorText?: React.ReactNode;
+  warningText?: React.ReactNode;
+  loading?: boolean;
+  readOnly?: boolean;
+  i18nStrings: FileTokenProps.I18nStrings;
+  dismissLabel?: string;
+  alignment?: TokenGroupProps.Alignment;
+  groupContainsImage?: boolean;
+  isImage: boolean;
+  index: number;
+}
+
+function InternalFileToken({
+  file,
+  showFileLastModified,
+  showFileSize,
+  showFileThumbnail,
+  i18nStrings,
+  onDismiss,
+  errorText,
+  warningText,
+  readOnly,
+  loading,
+  alignment,
+  groupContainsImage,
+  isImage,
+  index,
+}: FileTokenProps) {
+  const formatFileSize = i18nStrings.formatFileSize ?? defaultFormatters.formatFileSize;
+  const formatFileLastModified = i18nStrings.formatFileLastModified ?? defaultFormatters.formatFileLastModified;
+
+  const errorId = useUniqueId('error');
+  const warningId = useUniqueId('warning');
+
+  const showWarning = warningText && !errorText;
+  const containerRef = useRef<HTMLDivElement>(null);
+  const fileNameRef = useRef<HTMLSpanElement>(null);
+  const fileNameContainerRef = useRef<HTMLDivElement>(null);
+  const [showTooltip, setShowTooltip] = useState(false);
+
+  function isEllipsisActive() {
```

Comment:
We do. We need this to determine when to show the tooltip, so it needs to be in javascript. Not sure if renaming this "isTooltipActive" would be more apt, since this is literally what its doing?